File: test/Models/Buildable/Installable/HeaderTest.java
package Models.Buildable.Installable;

import Graphics.GraphicsList;
import Graphics.RectangleInstructions;
import Models.Buildable.Material.Lumber;
import Models.Buildable.Material.MaterialList;
import Models.Buildable.Material.Nail;
import Models.Measurement;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class HeaderTest {
    Lumber.Dimension dimension = Lumber.Dimension.TWO_BY_FOUR;
    @Test
    public void shouldReturnMaterialList() {
        Measurement width = new Measurement(41);
        MaterialList result = new MaterialList()
                .addMaterial(new Lumber(width, Lumber.Dimension.TWO_BY_FOUR), 2)
                .addMaterial(new Lumber(width, Lumber.Dimension.TWO_BY_SIX), 2)
                .addMaterial(Nail.TEN_D, 48);
        assertEquals(result, new Header(width).materialList());
    }

    @Test
    public void shouldReturnGraphicsList() {
        Measurement zero = new Measurement(0);
        Measurement gap = new Measurement(41);
        GraphicsList result = new GraphicsList()
                .addGraphic(new RectangleInstructions(zero, zero, gap, dimension.width))
                .addGraphic(new RectangleInstructions(zero, dimension.width, gap, Lumber.Dimension.TWO_BY_SIX.height))
                .addGraphic(new RectangleInstructions(zero, dimension.width.clone().add(Lumber.Dimension.TWO_BY_SIX.height), gap, Lumber.Dimension.TWO_BY_SIX.height))
                .addGraphic(new RectangleInstructions(zero, dimension.width.clone().add(Lumber.Dimension.TWO_BY_SIX.height.clone().multiply(2)), gap, dimension.width));
        assertEquals(result.drawingInstructions(), new Header(gap).graphicsList().drawingInstructions());
    }

    @Test
    public void shouldReturnTotalWidth() {
        Measurement gapWidth = new Measurement(44);
        assertEquals(gapWidth, new Header(gapWidth).totalWidth());
    }

    @Test
    public void shouldReturnTotalHeight() {
        Measurement result = new Measurement(8, Measurement.Fraction.ONE_HALF);
        assertEquals(result, new Header(new Measurement(41)).totalHeight());
    }

}

File: src/test/kotlin/models/buildable/installable/HeaderTest.kt
package models.buildable.installable

import graphics.*
import models.buildable.material.*
import models.Measurement
import models.Measurement.Fraction
import kotlin.test.*

internal class HeaderTest {
    private val dimension = Lumber.Dimension.TWO_BY_FOUR
    private val basicWidth = Measurement(44)
    private val basicHeader = Header(Stud(basicWidth, dimension), Plate(basicWidth, dimension))

    @Test
    fun shouldReturnMaterialList() {
        val studTest = Stud(basicWidth, Lumber.Dimension.TWO_BY_SIX)
        val plateTest = Plate(basicWidth, dimension)
        val result = MaterialList()
            .addMaterials(DoubleStud(studTest).materialList())
            .addMaterials(plateTest.materialList())
            .addMaterials(plateTest.materialList())
            .addMaterial(Nail.TEN_D, 12)
        assertEquals(result.materials(), Header(studTest, plateTest).materialList().materials())
    }

    @Test
    fun shouldReturnGraphicsList() {
        val zero = Measurement(0)
        val twoBySixHeight = Lumber.Dimension.TWO_BY_SIX.height
        val result = GraphicsList()
            .addGraphic(RectangleInstructions(zero, zero, basicWidth, dimension.width))
            .addGraphic(RectangleInstructions(zero, dimension.width, basicWidth, twoBySixHeight))
            .addGraphic(RectangleInstructions(zero, dimension.width.add(twoBySixHeight), basicWidth, dimension.width))
        assertEquals(
            result.drawingInstructions(),
            Header(Stud(basicWidth, Lumber.Dimension.TWO_BY_SIX), Plate(basicWidth, dimension))
                .graphicsList()
                .drawingInstructions()
        )
    }

    @Test
    fun shouldReturnTotalWidth() {
        assertEquals(basicWidth, basicHeader.totalWidth())
    }

    @Test
    fun shouldReturnTotalHeight() {
        assertEquals(Measurement(6, Fraction.ONE_HALF), basicHeader.totalHeight())
    }

    @Test
    fun whenPlateAndLoadBearingStudAreNotSameLengthShouldThrow() {
        val error = "LoadBearing stud and Plate must be the same length"
        val firstThrown = assertFailsWith<IllegalArgumentException> {
            Header(Stud(Measurement(10), dimension), Plate(Measurement(12), dimension))
        }
        assertEquals(error, firstThrown.message)
        val secondThrown = assertFailsWith<IllegalArgumentException> {
            Header(Stud(Measurement(14), dimension), Plate(Measurement(12), dimension))
        }
        assertEquals(error, secondThrown.message)
    }

}